### PR TITLE
Atmos changes

### PR DESCRIPTION
**Pull request by gradientvera**
Changes needed for atmos.
Some of these changes aren't needed anymore and I'll remove them in due time

### CODE REVIEW

**Comment by gradientvera** [comment]
Ready for review.

File: Robust.Shared.Maths/Direction.cs [hunk 1]
```diff
@@ -68,5 +95,15 @@ public static Angle ToAngle(this Vector2 vec)
         {
             return Math.Atan2(vec.Y, vec.X);
         }
+
+        /// <summary>
+        /// Converts a direction vector to an angle, where angle is -PI to +PI.
+        /// </summary>
+        /// <param name="vec">Vector to get the angle from.</param>
+        /// <returns>Angle of the vector.</returns>
+        public static Angle ToAngle(this Vector2i vec)
```
Comment (PJB3005) [inline Robust.Shared.Maths/Direction.cs]: You can already do this by doing `new Angle(vec)`

File: Robust.Shared/Map/Coordinates.cs [hunk 2]
```diff
@@ -97,6 +97,14 @@ public Vector2 ToMapPos(IMapManager mapManager)
             return mapManager.GetGrid(GridID).LocalToWorld(Position);
         }
 
+        /// <summary>
+        ///     Converts this set of coordinates to map indices.
+        /// </summary>
+        public MapIndices ToMapIndices()
+        {
+            return new MapIndices((int)MathF.Floor(X), (int)MathF.Floor(Y));
```
Comment (PJB3005) [inline Robust.Shared/Map/Coordinates.cs]: Grids are not guaranteed to have a tile size of one. This method is incorrect.

File: Robust.Shared/Map/MapIndices.cs [hunk 3]
```diff
@@ -83,6 +100,11 @@ public override string ToString()
             return $"{{{X},{Y}}}";
         }
 
+        public GridCoordinates ToGridCoordinates(GridId gridId)
+        {
+            return new GridCoordinates(X, Y, gridId);
```
Comment (PJB3005) [inline Robust.Shared/Map/MapIndices.cs]: Again, these are not necessarily correct.

File: Robust.Shared/Serialization/YamlObjectSerializer.cs [hunk 4]
```diff
@@ -602,8 +618,37 @@ public YamlNode TypeToNode(object obj)
                 return convertible.ToString(CultureInfo.InvariantCulture);
             }
 
-            // List<T>/IReadOnlyCollection<T>/IReadOnlyList<T>
-            if (TryGenericListType(type, out var listType) || TryGenericReadOnlyCollectionType(type, out listType))
+            // array
+            if (type.IsArray)
```
Comment (PJB3005) [inline Robust.Shared/Serialization/YamlObjectSerializer.cs:622]: You messed this merge up, and accidentally removed the `TryGenericReadOnlyCollectionType` code.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: Robust.Shared.Maths/Direction.cs (hunk 1 above)
Comment (gradientvera) [reply Robust.Shared.Maths/Direction.cs]: Done.

File: Robust.Shared/Map/Coordinates.cs (hunk 2 above)
Comment (gradientvera) [reply Robust.Shared/Map/Coordinates.cs]: Fixed this. Now it gets the actual tile indices from the grid.

File: Robust.Shared/Map/MapIndices.cs (hunk 3 above)
Comment (gradientvera) [reply Robust.Shared/Map/MapIndices.cs]: Changed this to multiply X and Y by the grid's tile size. I think that should be correct?

File: Robust.Shared/Serialization/YamlObjectSerializer.cs (hunk 4 above)
Comment (gradientvera) [reply Robust.Shared/Serialization/YamlObjectSerializer.cs:622]: Oops. Fixed now, I think?

**Comment by gradientvera** [comment]
Reviews addressed.

File: Robust.Shared/Serialization/YamlObjectSerializer.cs [hunk 5]
```diff
@@ -602,6 +618,35 @@ public YamlNode TypeToNode(object obj)
...
+                    }
+
+                    var entryNode = TypeToNode(entry);
+
+                    // write the concrete type tag
+                    if (element!.IsAbstract || element!.IsInterface)
+                    {
+                        var concreteType = entry.GetType();
+                        entryNode.Tag = $"!type:{concreteType.Name}";
+                    }
```
Comment (PJB3005) [inline Robust.Shared/Serialization/YamlObjectSerializer.cs]:
```suggestion
                    AssignTag<object?>(element, entry, null, entryNode);
```

**Review by PJB3005 — APPROVED** [review]

**Pull request merged**